packages/entity-server/src/routes/hierarchy/middleware/fields.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType, EntityFields } from '../../../models';
+import {
+  HierarchyRequest,
+  HierarchyContext,
+  ExtendedEntityFields,
+  EntityResponseObject,
+} from '../types';
+import {
+  extendedFieldFunctions,
+  getParentCodeBulk,
+  getChildrenCodesBulk,
+} from '../extendedFieldFunctions';
+
+const validFields: (keyof EntityFields)[] = ['code', 'country_code'];
+const isEntityField = (field: string): field is keyof EntityFields =>
+  (validFields as string[]).includes(field);
+const validateField = (field: string): field is keyof ExtendedEntityFields =>
+  isEntityField(field) || Object.keys(extendedFieldFunctions).includes(field);
+
+const allFields = [
+  ...validFields,
+  ...Object.keys(extendedFieldFunctions),
+] as (keyof ExtendedEntityFields)[];
+
+export const extractFieldsFromQuery = (queryFields?: string) => {
+  if (!queryFields) {
+    return allFields; // Display all fields if none specifically requested
+  }
+
+  const requestedFields = queryFields.split(',');
+  const fields = new Set<keyof ExtendedEntityFields>();
+  requestedFields.forEach(requestedField => {
+    if (validateField(requestedField)) {
+      fields.add(requestedField);
+    } else {
+      throw new Error(`Unknown field requested: ${requestedField}`);
```

Comment:
Will this error end up being returned as the query response? McLean has been promoting this idea of the "robustness principle" over in Tamanu, which is to be conservative with what you send but _liberal with what you accept_. That would suggest we should just ignore invalid fields here. I'm not totally sold on it when we control the clients as well as the server, but I'm interested to have it in mind as we watch people interact more with these micro-services

Reply:
Yes, it will return an error in the query response. The idea of tolerating 'invalid' fields is a bit of an interesting one with the way I've written the typescript here (where the keys of `extendedFieldFunctions` are a type that we use commonly). It could be handed, though I may need to do some refactoring...

As to the principal in general, I see the value but I dunno if see a strong reason to adopt it when as you say we are in control of the clients as well...

Might leave it as is for now?

Reply:
Yes sorry please do leave it as is for now! This one was more of a thought bubble cross-pollination idea than a suggested change for this PR :-)